File: windows/app/OpenBurnBar.App.MemorySearch/Memory/MemoryExtractionLlmClient.cs
using System.Threading;
using System.Threading.Tasks;

namespace OpenBurnBar.App.MemorySearch.Memory;

// PORTED (seam) from AgentLens/Services/Memory/MemoryExtractionLLMClient.swift.
//
// This is THE injectable network boundary for memory extraction. The Swift struct never throws —
// it returns null on any failure. HttpMemoryExtractionLlmClient implements the bounded HttpClient
// POST to an OpenAI-compatible /chat/completions or Ollama /api/generate; a fake returning canned
// JSON makes the whole extraction pipeline testable without a network. All cooldown/retry/progress
// bookkeeping lives in the CALLER (the transcript extractor), not in this seam.

/// <summary>Result of an Ollama completion. Swift: the <c>(text:shouldCooldown:)</c> tuple.</summary>
public readonly record struct OllamaCompletionResult(string? Text, bool ShouldCooldown);

/// <summary>The memory-extraction LLM boundary. Swift: <c>struct MemoryExtractionLLMClient</c>.</summary>
public interface IMemoryExtractionLlmClient
{
    /// <summary>
    /// POST to an OpenAI-compatible <c>/chat/completions</c>. Body: model, [system, user] messages,
    /// temperature 0.1, max_tokens, response_format json_object (and reasoning_effort "high" when
    /// the model id contains "gpt-5.5"). Returns the parsed content, or null on any failure.
    /// Swift: <c>callOpenAICompatibleCompletion(...)</c>.
    /// </summary>
    Task<string?> CallOpenAiCompatibleCompletionAsync(
        string baseUrl,
        string apiKey,
        string model,
        string systemPrompt,
        string userPrompt,
        double timeoutSeconds,
        int maxOutputTokens,
        bool includeOpenRouterHeaders,
        CancellationToken cancellationToken = default);

    /// <summary>
    /// POST to Ollama <c>/api/generate</c> (stream false, format json, options temperature 0.1 /
    /// num_predict). Returns the response text plus a cooldown hint (transport error, or status
    /// 404/408/429/&gt;=500). Swift: <c>callOllama(...)</c>.
    /// </summary>
    Task<OllamaCompletionResult> CallOllamaAsync(
        string baseUrl,
        string model,
        string systemPrompt,
        string userPrompt,
        double timeoutSeconds,
        int maxOutputTokens,
        CancellationToken cancellationToken = default);
}

File: AgentLens/Services/Memory/MemoryExtractionLLMClient.swift
import Foundation

// MARK: - Memory Extraction LLM Client
//
// Stateless, `Sendable` client for calling Ollama and OpenAI-compatible completion
// endpoints to extract durable memories. A direct sibling of `SummaryLLMClient`: same
// request/response shape, same error posture (returns `nil` on any failure, never
// throws), so it is safe to call from concurrent workers. All mutable cooldown /
// progress state stays in the caller (`ChatTranscriptExtractor`).
//
// The only material differences from `SummaryLLMClient` are the system prompt (memory
// extraction, not summarization) and an OpenAI `response_format: json_object` hint so
// compliant servers return strict JSON. The hint is best-effort; the parser tolerates
// prose-wrapped JSON either way.

/// Stateless request/response I/O for the memory-extraction LLM calls.
struct MemoryExtractionLLMClient: Sendable {

    // MARK: - OpenAI-Compatible Completion

    /// Calls an OpenAI-compatible `/chat/completions` endpoint and returns the
    /// assistant's reply text, or `nil` on any failure.
    func callOpenAICompatibleCompletion(
        baseURL: String,
        apiKey: String,
        model: String,
        systemPrompt: String,
        userPrompt: String,
        timeout: Double,
        maxOutputTokens: Int,
        includeOpenRouterHeaders: Bool
    ) async -> String? {
        let trimmedBase = baseURL.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let url = URL(string: trimmedBase + "/chat/completions") else { return nil }

        var request = URLRequest(url: url)
        request.timeoutInterval = timeout
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if apiKey.isEmpty == false {
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }
        if includeOpenRouterHeaders {
            request.setValue("OpenBurnBar", forHTTPHeaderField: "X-Title")
        }

        var body: [String: Any] = [
            "model": model,
            "messages": [
                ["role": "system", "content": systemPrompt],
                ["role": "user", "content": userPrompt]
            ],
            "temperature": 0.1,
            "max_tokens": maxOutputTokens,
            "response_format": ["type": "json_object"]
        ]
        if model.trimmingCharacters(in: .whitespacesAndNewlines).localizedCaseInsensitiveContains("gpt-5.5") {
            body["reasoning_effort"] = "high"
        }

        guard let requestBody = try? JSONSerialization.data(withJSONObject: body) else { return nil } // try?-ok(encode request, skip)
        request.httpBody = requestBody

        guard let (data, response) = try? await URLSession.shared.data(for: request), // try?-ok(network fetch, skip)
              let http = response as? HTTPURLResponse,
              (200 ..< 300).contains(http.statusCode)
        else {
            return nil
        }
        return OpenAICompatibleChatResponse.assistantText(fromResponseBody: data)
    }

    // MARK: - Ollama

    /// Calls a local Ollama `/api/generate` endpoint and returns the raw response
    /// text plus a cooldown hint, or `(nil, shouldCooldown)` on failure. Mirrors
    /// `SummaryLLMClient.callOllama` so cooldown bookkeeping stays in the caller.
    func callOllama(
        baseURL: String,
        model: String,
        systemPrompt: String,
        userPrompt: String,
        timeout: Double,
        maxOutputTokens: Int
    ) async -> (text: String?, shouldCooldown: Bool) {
        let base = baseURL.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let endpoint = URL(string: base)?.appendingPathComponent("api/generate"),
              model.isEmpty == false
        else {
            return (nil, false)
        }

        var request = URLRequest(url: endpoint)
        request.timeoutInterval = timeout
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let payload: [String: Any] = [
            "model": model,
            "system": systemPrompt,
            "prompt": userPrompt,
            "stream": false,
            "format": "json",
            "options": [
                "temperature": 0.1,
                "num_predict": maxOutputTokens
            ]
        ]

        guard let body = try? JSONSerialization.data(withJSONObject: payload) else { // try?-ok(encode request, skip)
            return (nil, false)
        }
        request.httpBody = body

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            let nsError = error as NSError
            let cooldown = nsError.domain == NSURLErrorDomain
            return (nil, cooldown)
        }

        guard let http = response as? HTTPURLResponse else { return (nil, false) }
        guard (200 ..< 300).contains(http.statusCode) else {
            let cooldown = http.statusCode == 404 || http.statusCode == 408
                || http.statusCode == 429 || http.statusCode >= 500
            return (nil, cooldown)
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any], // try?-ok(decode LLM JSON)
              let text = json["response"] as? String
        else {
            return (nil, false)
        }
        return (text, false)
    }
}
